Sources/MongoSwift/BSON/BsonValue.swift

```diff
@@ -277,51 +231,56 @@ extension Date: BsonValue {
         }
     }
 
-    public static func from(iter: inout bson_iter_t) -> BsonValue {
-        return Date(msSinceEpoch: bson_iter_date_time(&iter))
+    public init(from iter: DocumentIterator) throws {
+        self.init(msSinceEpoch: bson_iter_date_time(&iter.iter))
     }
 }
 
 /// An internal struct to represent the deprecated DBPointer type. While DBPointers cannot
 /// be created, we may need to parse them into `Document`s, and this provides a place for that logic.
 internal struct DBPointer: BsonValue {
 
-    public var bsonType: BsonType { return .dbPointer }
+    var bsonType: BsonType { return .dbPointer }
 
-    public func encode(to storage: DocumentStorage, forKey key: String) throws {
-        throw MongoError.bsonEncodeError(message: "DBPointers are deprecated; use a DBRef instead")
+    func encode(to storage: DocumentStorage, forKey key: String) throws {
+        throw MongoError.bsonEncodeError(message: "`DBPointer`s are deprecated; use a DBRef document instead")
     }
 
-    public static func from(iter: inout bson_iter_t) -> BsonValue {
+    init(from iter: DocumentIterator) throws {
+        throw MongoError.bsonDecodeError(message:
+            "`DBPointer`s are deprecated; use `DBPointer.asDocument` to create a DBRef document instead")
+    }
+
+    /// Reads DBPointer data from `iter` and converts it to DBRef format
+    static func asDocument(from iter: DocumentIterator) throws -> Document {
         var length: UInt32 = 0
         let collectionPP = UnsafeMutablePointer<UnsafePointer<Int8>?>.allocate(capacity: 1)
         defer {
             collectionPP.deinitialize(count: 1)
             collectionPP.deallocate(capacity: 1)
         }
+
         let oidPP = UnsafeMutablePointer<UnsafePointer<bson_oid_t>?>.allocate(capacity: 1)
         defer {
             oidPP.deinitialize(count: 1)
             oidPP.deallocate(capacity: 1)
         }
-        bson_iter_dbpointer(&iter, &length, collectionPP, oidPP)
 
-        guard let key = bson_iter_key(&iter) else {
-            preconditionFailure("Failed to retrieve key for DBPointer value")
-        }
+        bson_iter_dbpointer(&iter.iter, &length, collectionPP, oidPP)
+
         guard let oidP = oidPP.pointee else {
-            preconditionFailure(retrieveErrorMsg(type: "DBPointer ObjectId", key: String(cString: key)))
+            throw MongoError.bsonDecodeError(message:
+                retrieveErrorMsg(type: "DBPointer ObjectId", key: iter.currentKey))
         }
         guard let collectionP = collectionPP.pointee else {
-            preconditionFailure(retrieveErrorMsg(type: "DBPointer collection name", key: String(cString: key)))
+            throw MongoError.bsonDecodeError(message:
+                retrieveErrorMsg(type: "DBPointer collection name", key: iter.currentKey))
         }
 
-        let dbRef: Document = [
+        return [
```

Comment:
Since the return type is a Document, does Swift automatically initialize one from this returned dictionary literal?